### PR TITLE
Add attr 'axis' for softmax

### PR DESCRIPTION
**Pull request by heavengate**
屏幕快照 2019-03-09 下午5 21 11
屏幕快照 2019-03-09 下午5 21 22

### CODE REVIEW

File: paddle/fluid/operators/softmax_cudnn_op.cu.cc [hunk 1]
```diff
@@ -24,48 +25,96 @@ template <typename T>
...
     Out->mutable_data<T>(context.GetPlace());
 
-    auto dims = X->dims();
-    auto flattened_dims = framework::flatten_to_2d(dims, dims.size() - 1);
-    framework::LoDTensor flattened_x;
-    framework::LoDTensor flattened_out;
-    flattened_x.ShareDataWith(*X).Resize(flattened_dims);
-    flattened_out.ShareDataWith(*Out).Resize(flattened_dims);
+    std::vector<int> perm, shape;
+    CalcTransPermAndShapeByAxis(*X, axis, &perm, &shape);
```
Comment (qingqing01) [inline paddle/fluid/operators/softmax_cudnn_op.cu.cc]: This is not good for GPU.

Comment (heavengate) [reply paddle/fluid/operators/softmax_cudnn_op.cu.cc]: 修改为通过CPU/GPU kernel和jit kernel实现

File: paddle/fluid/operators/jit/more/mkl/mkl.cc [hunk 2]
```diff
@@ -144,6 +172,11 @@ bool VScalKernel<float>::CanBeUsed(const int& d) const {
   return platform::MayIUse(platform::avx512f) && d > 512;
 }
 
+template <>
+bool StrideScalKernel<float>::CanBeUsed(const int& d) const {
+  return platform::MayIUse(platform::avx512f) && d > 512;
```
Comment (tensor-tang) [inline paddle/fluid/operators/jit/more/mkl/mkl.cc]: use true.

File: paddle/fluid/operators/jit/more/mkl/mkl.cc [hunk 3]
```diff
@@ -128,6 +146,16 @@ void ASum<double>(const double* x, double* res, int n) {
   res[0] = platform::dynload::cblas_dasum(n, x, 1);
 }
 
+template <>
+void StrideSum<float>(const float* x, float* res, int n, int stride) {
+  res[0] = platform::dynload::cblas_sasum(n, x, stride);
```
Comment (tensor-tang) [inline paddle/fluid/operators/jit/more/mkl/mkl.cc]: asum=sum(abs(x)) is not stridesum

**Review by tensor-tang — COMMENTED** [review]
Please add two more unit tests for `StrideASum` and `StrideScal`

File: paddle/fluid/operators/jit/more/mkl/mkl.cc (hunk 3 above)
Comment (heavengate) [reply paddle/fluid/operators/jit/more/mkl/mkl.cc]: rename to StrideASum

File: paddle/fluid/operators/jit/more/mkl/mkl.cc (hunk 2 above)
Comment (heavengate) [reply paddle/fluid/operators/jit/more/mkl/mkl.cc]: Done

**Review by heavengate — COMMENTED** [review]
Add test for StrideASum and StrideScal

**Review by qingqing01 — DISMISSED** [review]

**Comment by heavengate** [comment]
@haowang101779990 @shanyi15  plz check doc

**Comment by shanyi15** [comment]
image
例子如果可以丰富下就更好了

File: paddle/fluid/operators/jit/kernel_base.h [hunk 4]
```diff
@@ -53,6 +53,8 @@ typedef enum {
   kVSquare,
   kVSub,
   kVTanh,
+  kStrideASum,
```
Comment (tensor-tang) [inline paddle/fluid/operators/jit/kernel_base.h]: https://github.com/PaddlePaddle/Paddle/pull/16057/files#diff-5f471713bbdd64a5b6631c94e80c518fR26

Please sort by alphabet

File: paddle/fluid/operators/jit/more/mix/mix.cc [hunk 5]
```diff
@@ -50,10 +50,14 @@ void VTanh(const T* x, T* y, int n) {
   compute_addbias(&b, y, y, n);
 }
 
-void Softmax(const T* x, T* y, int n, int bs) {
+void Softmax(const T* x, T* y, int n, int bs, int remain) {
   auto compute_hmax = KernelFuncs<HMaxTuple<T>, CPUPlace>::Cache().At(n);
   auto compute_hsum = KernelFuncs<HSumTuple<T>, CPUPlace>::Cache().At(n);
   auto compute_vscal = KernelFuncs<VScalTuple<T>, CPUPlace>::Cache().At(n);
+  auto compute_stridesum =
```
Comment (tensor-tang) [inline paddle/fluid/operators/jit/more/mix/mix.cc]: compute_stridesum => compute_strideasum

File: paddle/fluid/operators/jit/refer/refer.h [hunk 6]
```diff
@@ -411,19 +411,46 @@ void HSum(const T* x, T* res, int n) {
...
+      y[i] = x[i];
+    }
+  }
+}
+
 // y = e^(x - max(x))
 // y = y / sum(y)
 template <typename T>
-void Softmax(const T* x, T* y, int n, int bs = 1) {
+void Softmax(const T* x, T* y, int n, int bs = 1, int remain = 1) {
```
Comment (tensor-tang) [inline paddle/fluid/operators/jit/refer/refer.h:437]: please add some simple comments to tell what is `remain` ?

File: paddle/fluid/operators/jit/test.cc [hunk 7]
```diff
@@ -723,39 +723,124 @@ void TestKernelSoftmax() {
...
+    }
+  }
+}
+
+template <typename KernelTuple, typename PlaceType>
+void TestKernelStrideScal() {
+  using T = typename KernelTuple::data_type;
+  VLOG(10) << "Test JITKernel: " << jit::to_string(KernelTuple::kernel_type);
+  // for (int d : TestSizes()) {
+  //   for (int m : {1, 2, 3}) { // stride
```
Comment (tensor-tang) [inline paddle/fluid/operators/jit/test.cc]: update these.

**Comment by heavengate** [comment]
> image
> 例子如果可以丰富下就更好了

Done

File: paddle/fluid/operators/jit/kernel_base.h (hunk 4 above)
Comment (heavengate) [reply paddle/fluid/operators/jit/kernel_base.h]: Done

File: paddle/fluid/operators/jit/more/mix/mix.cc (hunk 5 above)
Comment (heavengate) [reply paddle/fluid/operators/jit/more/mix/mix.cc]: Done

File: paddle/fluid/operators/jit/refer/refer.h (hunk 6 above)
Comment (heavengate) [reply paddle/fluid/operators/jit/refer/refer.h:437]: Done

File: paddle/fluid/operators/jit/test.cc (hunk 7 above)
Comment (heavengate) [reply paddle/fluid/operators/jit/test.cc]: Done

**Comment by shanyi15** [comment]
最新的commit好像没有修文档，请再check下~

**Comment by heavengate** [comment]
> 最新的commit好像没有修文档，请再check下~

在倒数第二个commit里修了，https://github.com/PaddlePaddle/Paddle/pull/16057/commits/eb2123e12dc0ce1f6920aefa12b684f01bf9ca17

**Review by tensor-tang — APPROVED** [review]
LGTM about the softmax kernels

**Comment by qingqing01** [comment]
Need to add axis for softmax_with_cross_entropy in next PR.

**Review by shanyi15 — APPROVED** [review]
LG comments

**Review by XiaoguangHu01 — APPROVED** [review]

**Pull request merged**